### PR TITLE
Fix SAC loss explode

### LINKED ISSUE #332: SAC's loss explode on Hopper-v3
- [ ] I have marked all applicable categories:
    + [ ] exception-raising bug
    + [x] RL algorithm bug
    + [ ] documentation request (i.e. "X is missing from the documentation.")
    + [ ] new feature request
- [x] I have visited the [source website](https://github.com/thu-ml/tianshou/)
- [x] I have searched through the [issue tracker](https://github.com/thu-ml/tianshou/issues) for duplicates
- [x] I have mentioned version numbers, operating system and environment, where applicable:
  ```python
  import tianshou, torch, sys
  print(tianshou.__version__, torch.__version__, sys.version, sys.platform)
  ```

## Problems
I am trying to reproduce the results mentioned in the [mujoco benchmark](https://github.com/thu-ml/tianshou/tree/master/examples/mujoco/benchmark). And the command I used is `./run_experiments.sh Hopper-v3`, which runs SAC algorithm. The losses exploded after several epochs. I have repeated the experiments for 6 times and tried fixed alpha and auto alpha tuning, but all of them exploded finally.

## Other info
- Version: I tried 6426a39 and e605bde, both of them have this issue
- Operating system: ubuntu 18.04
- Environment: 0.4.0 1.8.1+cu102 3.7.10 (default, Feb 26 2021, 18:47:35)  [GCC 7.3.0] Linux

image
image

### PR DESCRIPTION
**Pull request by ChenDRAG**
fix #332

### CODE REVIEW

File: tianshou/policy/modelfree/ddpg.py [hunk 1]
```diff
@@ -55,6 +55,9 @@ def __init__(
     ) -> None:
         super().__init__(action_scaling=action_scaling,
                          action_bound_method=action_bound_method, **kwargs)
+        assert action_bound_method != "tanh", "tanh mapping is not supported" \
+            "in policies where action is used as input of critic , because" \
+            "raw action in range (-inf, inf) will cause instability in training"
```
Comment (Trinkle23897) [inline tianshou/policy/modelfree/ddpg.py:60]: this will cause error when SACPolicy passes in `action_bound_method="tanh"`

File: tianshou/policy/modelfree/sac.py [hunk 2]
```diff
@@ -63,13 +63,11 @@ def __init__(
...
         exploration_noise: Optional[BaseNoise] = None,
         deterministic_eval: bool = True,
-        action_bound_method: str = "tanh",
         **kwargs: Any,
     ) -> None:
         super().__init__(
             None, None, None, None, tau, gamma, exploration_noise,
-            reward_normalization, estimation_step,
-            action_bound_method=action_bound_method, **kwargs)
+            reward_normalization, estimation_step, **kwargs)
```
Comment (Trinkle23897) [inline tianshou/policy/modelfree/sac.py:70]: add `action_bound_method="tanh"`? But this will cause DiscreteSACPolicy init error (because it needs `action_bound_method=""`

File: tianshou/policy/modelfree/td3.py [hunk 3]
```diff
@@ -35,8 +35,8 @@ class TD3Policy(DDPGPolicy):
     :param bool action_scaling: whether to map actions from range [-1, 1] to range
         [action_spaces.low, action_spaces.high]. Default to True.
     :param str action_bound_method: method to bound action to range [-1, 1], can be
-        either "clip" (for simply clipping the action), "tanh" (for applying tanh
-        squashing) for now, or empty string for no bounding. Default to "clip".
+        either "clip" (for simply clipping the action) or empty string for no bounding.
+        Default to "tanh".
```
Comment (Trinkle23897) [inline tianshou/policy/modelfree/td3.py]:
```suggestion
        Default to "clip".
```

File: tianshou/policy/modelfree/sac.py [hunk 4]
```diff
@@ -120,20 +118,20 @@ def forward(  # type: ignore
...
+        if self.action_scaling and self.action_space is not None:
+            action_scale = to_torch_as(
+                (self.action_space.high - self.action_space.low) / 2.0, act)
+        else:
+            action_scale = 1.0  # type: ignore
+        squashed_action = torch.tanh(act)
+        log_prob = log_prob - torch.log(
+            action_scale * (1 - squashed_action.pow(2)) + self.__eps
+        ).sum(-1, keepdim=True)
+        return Batch(logits=logits, act=squashed_action,
```
Comment (Trinkle23897) [inline tianshou/policy/modelfree/sac.py:133]: `act=squashed_action` will cause error in larger action space since you abandon the map_action function:
```python
In [1]: import gym

In [2]: e=gym.make("Pendulum-v0")

In [3]: e.action_space.high
Out[3]: array([2.], dtype=float32)

In [4]: e.action_space.low
Out[4]: array([-2.], dtype=float32)
```

Comment (ChenDRAG) [reply tianshou/policy/modelfree/sac.py:133]: In sac, this is solved by action scaling in map action. But for imitation policy, I think you are right.
`    il_policy.map_action = policy.map_action`
is indeed necessary.

**Comment by ChenDRAG** [comment]
Could you give some opinion, please? @danagi

**Review by danagi — DISMISSED** [review]

**Review by Trinkle23897 — APPROVED** [review]
